src/main/scala/example/Task1RepartitioningApproach.scala

```diff
@@ -0,0 +1,85 @@
+package example
+
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.apache.spark.sql.functions.col
+
+object Task1RepartitioningApproach {
+  // QUERY: Find spendings (both already paid or not) of different users on products from different categories.
+
+  def transformationTask1WithRepartitioningApproach(
+      spark: SparkSession,
+      notCompletedOrdersDF: DataFrame,
+      completedOrdersDF: DataFrame,
+      notCompletedPaymentsDF: DataFrame,
+      completedPaymentsDF: DataFrame,
+      usersDF: DataFrame,
+      productsDF: DataFrame,
+      categoriesDF: DataFrame
+    ): DataFrame = {
+    usersDF
+      .repartition(col("UserId"))
+      .createOrReplaceTempView("users")
+    productsDF
+      .repartition(col("CategoryId"))
+      .createOrReplaceTempView("products")
+    categoriesDF
+      .repartition(col("CategoryId"))
+      .createOrReplaceTempView("categories")
+//  completedPaymentsDF
+//    .createOrReplaceTempView("completed_payments")
+//  notCompletedPaymentsDF
+//    .createOrReplaceTempView("not_completed_payments")
+//  completedOrdersDF
+//    .createOrReplaceTempView("completed_orders")
+//  notCompletedOrdersDF
+//    .createOrReplaceTempView("not_completed_orders")
+    completedOrdersDF
+      .repartition(col("UserId"))
+      .union(notCompletedOrdersDF
+        .repartition(col("UserId")))
+      .createOrReplaceTempView("orders")
+
+    val groupedByUsersAndCategoriesDF = spark.sql("SELECT users.UserId AS ConsideredUserId, categories.CategoryId AS ConsideredCategoryId, SUM(orders.TotalValue) AS TotalSum FROM users LEFT JOIN orders ON users.UserId = orders.UserId LEFT JOIN products ON orders.ProductId = products.ProductId LEFT JOIN categories ON products.CategoryId = categories.CategoryId GROUP BY users.UserId, categories.CategoryId ORDER BY ConsideredUserId ASC, TotalSum DESC")
+
+    // groupedByUsersAndCategoriesDF.explain()
+
+    //  == Physical Plan ==
+    //  AdaptiveSparkPlan isFinalPlan=false
```

Comment:
Ok, nice. What can you learn from the plan?

Reply:
I've commented some of the execution plans, what happens at different points of them. Not all of them since it's a repetitive task. We can see that we are doing some additional repartitioning at the beginning, but the rest of the later tasks is a bit faster later in Spark UI.